src/CoreWCF.Primitives/src/CoreWCF/IdentityModel/Selectors/SamlSecurityTokenAuthenticator.cs

```diff
@@ -0,0 +1,243 @@
+        {
+            get { return _allowedAudienceUris; }
+        }
+
+        protected override bool CanValidateTokenCore(SecurityToken token)
+        {
+            return token is SamlSecurityToken;
+        }
+
+        protected override ValueTask<ReadOnlyCollection<IAuthorizationPolicy>> ValidateTokenCoreAsync(Tokens.SecurityToken token)
+        {
+            if (token == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperArgumentNull(nameof(token));
+
+            SamlSecurityToken samlToken = token as SamlSecurityToken;
+
+            if (samlToken == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlTokenAuthenticatorCanOnlyProcessSamlTokens, token.GetType().ToString())));
+
+            if (samlToken.Assertion.Signature == null)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlTokenMissingSignature)));
+
+            if (!IsCurrentlyTimeEffective(samlToken))
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLTokenTimeInvalid, DateTime.UtcNow.ToUniversalTime(), samlToken.ValidFrom.ToString(), samlToken.ValidTo.ToString())));
+
+          //  if (samlToken.Assertion.SigningToken == null)
+          //      throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlSigningTokenMissing)));
+
+            // Build the Issuer ClaimSet for this Saml token.
+            ClaimSet issuer = null;
+            bool canBeValidated = false;
+            for (int i = 0; i < _supportingAuthenticators.Count; ++i)
+            {
+                canBeValidated = _supportingAuthenticators[i].CanValidateToken(samlToken);
+                if (canBeValidated)
+                    break;
+            }
+            if (!canBeValidated)
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SamlInvalidSigningToken)));
+            ReadOnlyCollection<IAuthorizationPolicy> policies;
+            issuer = ResolveClaimSet(samlToken, out policies) ?? ClaimSet.Anonymous;
+
+            if ((_audienceUriMode == AudienceUriMode.Always)
+             || (_audienceUriMode == AudienceUriMode.BearerKeyOnly) && (samlToken.SecurityKeys.Count < 1))
+            {
+                // throws if not found.
+                bool foundAudienceCondition = false;
+                if (_allowedAudienceUris == null)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new SecurityTokenException(SR.Format(SR.SAMLAudienceUrisNotFound)));
+                }
+
+                foreach (MSSaml.SamlCondition samlCondition in  samlToken.Assertion.Conditions.Conditions)
+                {
+                    if(samlCondition is MSSaml.SamlAudienceRestrictionCondition)
+                    {
```

Comment:
You can do this to avoid an is check and a cast:
```c#
if (samlCondition is MSSaml.SamlAudienceRestrictionCondition audienceCondition)
{
    foundAudienceCondition = true;
    if (!ValidateAudienceRestriction(audienceCondition))
    {
....
```
It's one of the new(-ish) features in C#